File: src/phorge/api/endpoints/daemons.py
"""Daemon API endpoints."""

from __future__ import annotations

from typing import Any

from phorge.api.client import ForgeClient
from phorge.api.models import Daemon


class DaemonsAPI:
    def __init__(self, client: ForgeClient) -> None:
        self._client = client

    async def list(self, server_id: int) -> list[Daemon]:
        data = await self._client.get(f"/servers/{server_id}/daemons")
        return [Daemon(**d) for d in data.get("daemons", [])]

    async def get(self, server_id: int, daemon_id: int) -> Daemon:
        data = await self._client.get(f"/servers/{server_id}/daemons/{daemon_id}")
        return Daemon(**data["daemon"])

    async def create(
        self,
        server_id: int,
        *,
        command: str,
        user: str = "forge",
        directory: str | None = None,
        processes: int = 1,
        startsecs: int = 1,
    ) -> Daemon:
        payload: dict[str, Any] = {
            "command": command,
            "user": user,
            "processes": processes,
            "startsecs": startsecs,
        }
        if directory:
            payload["directory"] = directory
        data = await self._client.post(f"/servers/{server_id}/daemons", json=payload)
        return Daemon(**data["daemon"])

    async def restart(self, server_id: int, daemon_id: int) -> None:
        await self._client.post(f"/servers/{server_id}/daemons/{daemon_id}/restart")

    async def delete(self, server_id: int, daemon_id: int) -> None:
        await self._client.delete(f"/servers/{server_id}/daemons/{daemon_id}")

File: internal/forge/daemons.go
package forge

import (
	"context"
	"net/http"
)

// DaemonCreateOpts contains the options for creating a daemon.
type DaemonCreateOpts struct {
	Command   string `json:"command"`
	User      string `json:"user"`                // default "forge"
	Directory string `json:"directory,omitempty"` // optional
	Processes int    `json:"processes"`           // default 1
}

// List returns all background processes on a server.
func (s *DaemonsService) List(ctx context.Context, serverID int64) ([]BackgroundProcess, error) {
	return listResources[BackgroundProcess](s.client, ctx, s.client.orgPath("/servers/%d/background-processes", serverID))
}

// Get returns a single background process by ID.
func (s *DaemonsService) Get(ctx context.Context, serverID, daemonID int64) (*BackgroundProcess, error) {
	return getResource[BackgroundProcess](s.client, ctx, s.client.orgPath("/servers/%d/background-processes/%d", serverID, daemonID))
}

// Create creates a new background process on a server.
func (s *DaemonsService) Create(ctx context.Context, serverID int64, opts DaemonCreateOpts) (*BackgroundProcess, error) {
	return createResource[BackgroundProcess](s.client, ctx, s.client.orgPath("/servers/%d/background-processes", serverID), opts)
}

// Restart restarts a background process.
func (s *DaemonsService) Restart(ctx context.Context, serverID, daemonID int64) error {
	path := s.client.orgPath("/servers/%d/background-processes/%d/actions", serverID, daemonID)
	body := map[string]string{"action": "restart"}
	return s.client.do(ctx, http.MethodPost, path, body, nil)
}

// Delete removes a background process from a server.
func (s *DaemonsService) Delete(ctx context.Context, serverID, daemonID int64) error {
	return s.client.do(ctx, http.MethodDelete, s.client.orgPath("/servers/%d/background-processes/%d", serverID, daemonID), nil, nil)
}
